### ISSUE TITLE
Routing in a Causal cluster

### ISSUE THREAD

**Issue by luiztucillo** [neo4j-php/Bolt#42]
Did this lib work with routing in a causal cluster?

When I connect and read data it's works, but writing get me an error for trying to write on non leader node.

There's a way to use that?

**Comment by stefanak-michal** [comment]
Method hello (init) accepts routing as forth argument. More about this argument is described here: https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---41---hello

**Comment by u12206050** [comment]
I am having the same problem and setting routing to [] doesn't work.
I am using Neo4j Aura

**Comment by stefanak-michal** [comment]
> I am having the same problem and setting routing to [] doesn't work.
> I am using Neo4j Aura

Hi. It will be nice if you send your piece of code.

**Comment by u12206050** [comment]
Here is an extraction of my code
Setup code for my neo4j class
```
protected function bolt()
{
            $ip = config('services.neo4j.ip');
            $port = config('services.neo4j.port');
            $version = config('services.neo4j.version');
            $username = config('services.neo4j.username');
            $password = config('services.neo4j.password');

            // Create new Bolt instance
            $conn = new StreamSocket($ip, $port);
            $conn->setSslContextOptions([
                'verify_peer' => true,
                'allow_self_signed' => true
            ]);
            $bolt = new Bolt($conn);

            // Set Bolt protocol version if needed
            /*if (in_array($version, [3, 4, 4.1])) {
                $bolt->setProtocolVersions($version);
            }*/

            // Connect to database
            $bolt->hello('Laravel/1.0', $username, $password, []);

            return $bolt;
}

public function execute(string $statement)
{
        $bolt = $this->bolt();
        $mode = preg_match('/(CREATE|SET|MERGE|DELETE)/m', $statement) ? 'w' : 'r';

        if (! $res = $bolt->run($statement, [], [
            'mode' => $mode
        ])) {
            return null;
        }

        $records = $bolt->pull();
        array_pop($records);

        return array_map(function ($record) {
                return $record[0];
        }, $records);
}
``` 

Example of usage:

```
$userId = 1; /* Logged in user id */
$postId = 10; /* Id of post */

$neo4j = new Neo4j();
$neo4j->execute("MERGE (u:User { id: $userId })-[r:READS]->(p:Post { id: $postId }) SET r.ts = timestamp() RETURN p.id");
```

**Comment by u12206050** [comment]
In the cluster I see there are three servers, two are FOLLOWERs and only one is the LEADER. 
I can run `CALL dbms.cluster.overview()` to view their details.

**Comment by stefanak-michal** [comment]
@u12206050 It looks okay for me. What version of Neo4j are you using?

**Comment by stefanak-michal** [comment]
@u12206050 also did you try set directly the address for routing leader? `["address": "your-address:port"]`

**Comment by u12206050** [comment]
Yes that is what I am currently doing although I am setting it as the main ip.

**Comment by u12206050** [comment]
Setting `["address" => "your-address:port"]` doesn't work, it only works when I put the address to the LEADER as the main ip

**Comment by u12206050** [comment]
Sorry, using version 4

**Comment by luiztucillo** [comment]
I opened this "issue" because I can't make the routing to work.
I use a causal cluster running in AWS with kubernetes and the official neo4j helm.
On some research i did i found that the routing implementation is in the language driver side not in the DBMS or protocol itself.
I get an example in the [go lang driver](https://github.com/neo4j/neo4j-go-driver) and some other oficial drivers.
In the end, i did something [like that](https://github.com/stefanak-michal/Bolt/issues/18#issuecomment-710081714) with some changes and adjustments for my needs, like caching and retries.
It'll be great if we have this in the driver side.
In other hand I don't think that in this case where the lib is bolt focused, not like an "ORM", it has this routing responsability.
That's just my opinion.
If you guys want to implement something like that or want to discuss about that, call me. =)

**Comment by u12206050** [comment]
Thanks @luiztucillo I was starting to implement something like that similar.
Which brings me to the question what does the routing argument in this library eg. `["address" => "your-address:port"]` actually do then, since there is also multiple addresses within a cluster?

**Comment by stefanak-michal** [comment]
Sadly I don't have Aura cluster to test it. I followed official documentation, you should connect to leader. I created local cluster at my pc and it works for me. After cluster runs I checked which one is leader and discovered it runs at 7689 port (neo4j database).

|  | id | addresses | databases | groups |
| -- | -- | -- | -- | -- |
| 1 | "dea18832-edf1-434d-bc13-3fd554c32925" | ["bolt://localhost:7687", "http://localhost:7474"] | {   "neo4j": "FOLLOWER",   "system": "LEADER" } | [] |
| 2 | "f42cb101-1f97-40ad-8ec8-8a1d21e7f71a" | ["bolt://localhost:7688", "http://localhost:7475"] | {   "neo4j": "FOLLOWER",   "system": "FOLLOWER" } | [] |
| 3 | "6bba34e1-3744-4399-a4ac-06f0a303dcf5" | ["bolt://localhost:7689", "http://localhost:7476"] | {   "neo4j": "LEADER",   "system": "FOLLOWER" } | [] |



I'm not sure about the logic which @luiztucillo mentioned, because the guy said, he is using Neo4j version 3.

This piece of code works for me without problem.

```php
$conn = new \Bolt\connection\Socket('127.0.0.1', 7689, 3);
$bolt = new \Bolt\Bolt($conn);
$bolt->init('MyClient/1.0', $user, $password, []);
$res = $bolt->run('RETURN 1 as num, 2 as cnt', [], ['mode' => 'r']);
$res = $bolt->run('CREATE (a:Test) RETURN a, ID(a)');
```

Also you didn't send what error are you receiving.

**Comment by Dreiser** [comment]
**TL;DR:** It's also not working for me (AWS EKS + Neo4j Helm chart)

I've set up the AWS EKS cluster and deployed the Neo4j helm chart there. At first, I was able to connect to the cluster and write to it but that was probably just because I got lucky and I got the right IP from the DNS record. Now with the same solution, I'm getting "_No write operations are allowed directly on this database. Writes must pass through the leader. The role of this server is: FOLLOWER (Neo.ClientError.Cluster.NotALeader)_"

My connection:
```
$bolt = new LibraryBolt(new StreamSocket('my-neo4j.default.svc.cluster.local', 7687));
$bolt->setProtocolVersions(4.2);
$bolt->init('Client/1.0', $username, $password, []);
```
@stefanak-michal I think that problem with your example is that you checked routing yourself and then connected to the right leader manually (by using the right port) rather than connect to the follower and let the server switch you with routing = [] (that example would work even without routing argument). The difference may also be in using DNS records vs connecting to a specific server/port (not sure, just putting it out here).

**Comment by stefanak-michal** [comment]
@Dreiser you are right. I've made check of port before connecting. 

I'll postpone this issue to Neo4j people to discover some solution.

**Comment by stefanak-michal** [comment]
### Important

I've response from Neo4j people (it was fast). Currently there is no solution for how to receive routing table or let Neo4j database to do it by itself. Application has to acquire routing info in another way and use it for connection. 

The Neo4j team is preparing change for this this year. Connection will be extended and they will add new message type to acquire routing info which can be used for connection message.

_I'll be informed and I'll try to implement it asap._

**Comment by u12206050** [comment]
That doesn't make sense though, i was able to run a query to get the
available servers (read, write) and the based on the type of query I want
to run I setup a connection to the correct server.

I only recheck the servers once a write query fails, possibly because
another server has become the new write server.

**Comment by stefanak-michal** [comment]
@u12206050 if you connect to follower, request that info, discover which one is leader, disconnect and connect again to leader, it should work. 

What query for that are you using?

**Comment by u12206050** [comment]
Yep something like that, here is my `Neo4jClient`

```

<?php


namespace App\Neo4j;

use Bolt\Bolt;
use Bolt\connection\StreamSocket;

class BoltClient
{
    protected $servers;
    protected $writer;
    protected $reader;

    private $ip;
    private $port;
    private $username;
    private $password;

    public function __construct ()
    {
        $this->ip = config('services.neo4j.ip');
        $this->port = config('services.neo4j.port');
        $this->username = config('services.neo4j.username');
        $this->password = config('services.neo4j.password');

        if (config('services.neo4j.cluster')) {
            $this->loadRoutingTable();
        } else {
            $this->reader = $this->writer =
$this->newConnection($this->ip, $this->port);
        }
    }

    protected function loadRoutingTable ()
    {
            return cache()->remember('neo4j-routing-table', 86400, function () {
                $bolt = $this->newConnection($this->ip, $this->port);
                $bolt->run('CALL dbms.routing.getRoutingTable({})');
                $rows = $bolt->pull();

                $writeServers = [];
                $readServers = [];

                if(isset($rows[0][1])) {
                    foreach($rows[0][1] as $server) {
                        if($server['role'] === 'WRITE') {
                            $writeServers = $server['addresses'];
                        } elseif($server['role'] === 'READ') {
                            $readServers = $server['addresses'];
                        }
                    }
                }

                return [
                    'w' => $writeServers,
                    'r' => $readServers
                ];
            });
    }

    protected function newConnection ($ip, $port)
    {
        // Create new Bolt instance
        $conn = new StreamSocket($ip, $port);
        $conn->setSslContextOptions([
            /*'peer_name' => 'localhost',*/
            'verify_peer' => true,
            'allow_self_signed' => true
        ]);
        $bolt = new Bolt($conn);

        // Connect to database
        $bolt->hello('Laravel/1.0', $this->username, $this->password, []);

        return $bolt;
    }

    public function getClient ($mode = 'w')
    {
        if ($mode === 'r') {
            return $this->getReader();
        }

        return $this->getWriter();
    }

    public function getWriter ()
    {
        if ($this->writer) {
            return $this->writer;
        }

        $writers = $this->servers['w'];
        if ($len = count($writers)) {
            $address = $len > 1 ? $writers[random_int(0, $len - 1)] :
$writers[0];
            $this->writer = $this->newConnection(...explode(':', $address));
            return $this->writer;
        }

        throw new \Exception('No NEO4J server loaded');
    }

    public function getReader ()
    {
        if ($this->reader) {
            return $this->reader;
        }

        $readers = $this->servers['r'];
        if ($len = count($readers)) {
            $address = $len > 1 ?$readers[random_int(0, $len - 1)] :
$readers[0];
            return $this->reader =
$this->newConnection(...explode(':', $address));
        }

        return $this->reader = $this->getWriter();
    }

    public function clearCache()
    {
        cache()->delete('neo4j-routing-table');

        $this->reader = $this->writer = null;
        $this->loadRoutingTable();
    }
}

```

**Comment by Dreiser** [comment]
Neo4j version 4.2.2 has a server-side routing option and with that enabled this driver works as expected.

**Comment by hidabe** [comment]
@Dreiser thanks, it is great solution, but i not found this information in changelog: https://github.com/neo4j/neo4j-browser/wiki/changelog, too, neo4j Aura has 4.2.0 driver

**Comment by hidabe** [comment]
hi again, i think that the decision about what server from cluster use is from neo-client, not neo-bolt, i have added this feature to my connection, similar to the @u12206050 user: https://github.com/hidabe/neo4j-php-client/blob/master/src/Connection/Connection.php it work using "bolt+routing" in schema

**Comment by Dreiser** [comment]
@hidabe Sorry for the confusion - I'm not sure when did that option got to neo4j I just mentioned the version I updated on where I switched on server-side routing.

Anyways, this driver should ideally still implement bolr+routing protocol because it could optimize connections directly on the client side and it could handle the faulty connection (using a different connection from the pool).

Right now when I'm connected to Neo4j causal cluster using DNS with multiple records and when the first record goes down, PHP fails to connection initialization with "Read error" and the script dies even though other members are still up. This resolves itself either when the member comes back up or when the DNS translates to another working host. But obviously, it kills high availability and does not take full advantage of routing on the client.

**Comment by fbiville** [comment]
For anyone stumbling upon this thread, https://github.com/neo4j-php/neo4j-php-client (developed by @transistive on top of this library) supports clusters and routing, so you should give it a try.

**Comment by stefanak-michal** [comment]
The Route message from latest neo4j version is implemented. But I know @transistive had to deal with some issues while trying to make it work. Maybe he has more insight for it.

https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---43---route

https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---44---route

https://github.com/neo4j-php/Bolt/blob/9dabf7de69c3e08270cdf26d0489e4a7dcb4c495/src/Bolt.php#L396

**Comment by transistive** [comment]
Hi All!

The client does indeed fully support clusters and routing. You can, of course, try and implement it yourself, but it is more challenging than it appears!

Good luck all :+1:
